pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/aws/karpenter/pkg/apis/settings"
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	awserrors "github.com/aws/karpenter/pkg/errors"
+)
+
+var (
+	instanceStateFilter = &ec2.Filter{
+		Name:   aws.String("instance-state-name"),
+		Values: aws.StringSlice([]string{ec2.InstanceStateNamePending, ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping, ec2.InstanceStateNameStopped, ec2.InstanceStateNameShuttingDown}),
+	}
+)
+
+type Provider struct {
+	region string
+	iamapi iamiface.IAMAPI
+	ec2api ec2iface.EC2API
+	cache  *cache.Cache
+}
+
+func NewProvider(region string, iamapi iamiface.IAMAPI, ec2api ec2iface.EC2API, cache *cache.Cache) *Provider {
+	return &Provider{
+		region: region,
+		iamapi: iamapi,
+		ec2api: ec2api,
+		cache:  cache,
+	}
+}
+
+func (p *Provider) Create(ctx context.Context, nodeClass *v1beta1.EC2NodeClass, tags map[string]string) (string, error) {
+	localTags := lo.Assign(tags, map[string]string{v1beta1.LabelNodeClass: nodeClass.Name, v1.LabelTopologyRegion: p.region})
+	profileName := p.GetProfileName(ctx, nodeClass)
+	delete(localTags, corev1beta1.NodePoolLabelKey)
+
+	// An instance profile exists for this NodeClass
+	if _, ok := p.cache.Get(string(nodeClass.UID)); ok {
+		return profileName, nil
+	}
+	// Validate if the instance profile exists and has the correct role assigned to it
+	var instanceProfile *iam.InstanceProfile
+	out, err := p.iamapi.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{InstanceProfileName: aws.String(profileName)})
+	if err != nil {
+		if !awserrors.IsNotFound(err) {
+			return "", fmt.Errorf("getting instance profile %q, %w", profileName, err)
+		}
+		o, err := p.iamapi.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
+			InstanceProfileName: aws.String(profileName),
+			Tags:                lo.MapToSlice(tags, func(k, v string) *iam.Tag { return &iam.Tag{Key: aws.String(k), Value: aws.String(v)} }),
+		})
+		if err != nil {
+			return "", fmt.Errorf("creating instance profile %q, %w", profileName, err)
+		}
+		instanceProfile = o.InstanceProfile
+	} else {
+		instanceProfile = out.InstanceProfile
+	}
+	if len(instanceProfile.Roles) == 1 {
+		return profileName, nil
+	}
+	if _, err = p.iamapi.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(profileName),
+		RoleName:            aws.String(nodeClass.Spec.Role),
+	}); err != nil {
+		return "", fmt.Errorf("adding role %q to instance profile %q, %w", nodeClass.Spec.Role, profileName, err)
+	}
+	p.cache.SetDefault(string(nodeClass.UID), nil)
+	return profileName, nil
+}
+
+func (p *Provider) AssociatedInstances(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]string, error) {
+	profileName := p.GetProfileName(ctx, nodeClass)
+
+	// Get all instances that are using our instance profile name and are not yet terminated
+	var ids []string
+	if err := p.ec2api.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
```

Comment:
QQ: does this automatically get included into our batcher?

Reply:
Good callout 👍 I'm not sure the batcher would really have a tangible impact here but I'm fine going down the same codepath to be more DRY.